### PR TITLE
Eliminate worker-to-worker communication.

### PR DESCRIPTION
**Pull request by ryzhyk**
TODOs:

- [ ] Handle channel disconnects in both client and worker code
- [x] Run benchmarks
- [x] Rebase on latest master
- [x] Changelog

### CODE REVIEW

File: rust/template/differential_datalog/src/program/mod.rs [hunk 1]
```diff
@@ -2262,20 +2280,36 @@ impl RunningProgram {
...
-                Ok(msg) => Err(format!(
-                    "received unexpected reply to flush request: {:?}",
-                    msg,
-                )),
-            }
+            self.waitfor_flush_ack()
         })
     }
+
+    fn waitfor_flush_ack(&self) -> Response<()> {
```
Comment (Kixiron) [inline rust/template/differential_datalog/src/program/mod.rs]: Total bikeshed, but I think I prefer a name like `.await_flush()` or `.wait_for_flush()`

File: rust/template/differential_datalog/src/program/worker.rs [hunk 2]
```diff
@@ -136,204 +94,138 @@ impl<'a> DDlogWorker<'a> {
...
+            }
 
-                    Err(TryRecvError::Empty) => {
-                        // Command channel empty: use idle time to work on garbage collection.
-                        // This will block when there is no more compaction left to do.
-                        // The sender must unpark worker 0 after sending to the channel.
-                        self.worker.step_or_park(None);
-                    }
+            // After each command received we step, if there's no timely work to be done
+            // our thread will be parked until it's re-awoken by a command
```
Comment (Kixiron) [inline rust/template/differential_datalog/src/program/worker.rs]:
```suggestion
            // After each batch of commands received we step, if there's no timely work to be done
            // our thread will be parked until it's re-awoken by a command
```

**Comment by ryzhyk** [comment]
The `stream` test is still timing out in CI, and it turns out I did not actually reproduce it.  Not sure if it's stuck or just slow.  Will continue investigating, but not sure how much progress I will make today.

**Comment by ryzhyk** [comment]
> The `stream` test is still timing out in CI, and it turns out I did not actually reproduce it. Not sure if it's stuck or just slow. Will continue investigating, but not sure how much progress I will make today.

Correction: the test completed eventually, but it took nearly 50 minutes, which does not seem right.

**Comment by ryzhyk** [comment]
I think I nailed down the performance issue. The reason we were not able to reproduce it is that it only shows up when there are many more workers than CPU cores available to DDlog.  Since this particular CI job runs on a public gitlab runner with a single virtual core, it was hit hard even though it only uses 4 workers.  I was able to reproduce this on my laptop using the `taskset` command on Linux to give a single CPU core to DDlog, causing the test to go 100x slower.

The bottleneck is in the crossbeam channel that uses a [spinlock](https://github.com/crossbeam-rs/crossbeam/blob/b71b94412cb1915a0cb2d4ffe043c67b7d935b8a/crossbeam-channel/src/waker.rs#L180) internally. User-level spinlocks are [fundamentally flawed](https://www.realworldtech.com/forum/?threadid=189711&curpostid=189723) and are virtually never a good idea, exactly because they cause this type of superlinear slowdown.  I am shocked that a serious library like crossbeam uses spinlocks. Specifically, several of the channels we use in DDlog are contended: worker-to-main thread channels, profiling channel, and the update handler channel.

Granted, this is a pathological scenario that issues 100K trivial transactions, so its runtime ends up being dominated by crossbeam messaging. But I still find the 100x slowdown disturbing.

I haven't yet investigated if this is a known problem with crossbeam and whether there exist any workarounds.  Will look into this now.  Other takeaways:

- Having an option to disable profiling altogether (or to enable it on demand) is beginning to sound like a good idea to me.
- It would be nice to try building a concurrent update handler, e.g., using dashmap to avoid relying on a global lock or a dedicated thread.

**Comment by Kixiron** [comment]
The spinlock is very unsettling, I would have thought crossbeam would use a more intelligent locking strategy (it'd be interesting to see if swapping the spinlock with something like parking_lot's mutex makes a difference). I definitely agree that being able to disable profiling would be a good thing, and I also have a few ideas for leveraging timely in order to make that more efficient, I'll play with it later. For the handling of updates I totally agree, I think actually leveraging timely would be a really great option as opposed to the current strategy, and if done right we could integrate it with channels and even `UnorderedInput`s to allow for completely concurrent transactions, also something I'll play with later

**Comment by Kixiron** [comment]
There's also multiple [exponential spinloops](https://github.com/crossbeam-rs/crossbeam/blob/b71b94412cb1915a0cb2d4ffe043c67b7d935b8a/crossbeam-utils/src/backoff.rs), this could be an incredibly pathological case of each thread backing off for increasingly longer periods

**Comment by ryzhyk** [comment]
Ok, my assessment was wrong. Spinlocks are silly and show up a lot in the `perf` profile (and we do need better update handler and profiler architecture).  But after a lot more experimentation, it appears that the real problem in this case is in the timely scheduler and it only shows up when the number of workers is larger than the number of cores and also depends on the number of opens session handles.  Here is what I did:

1. Simplified the DDlog program to a bare minimum input -> map -> output. No joins or any other interesting stuff going on. 
2. Disabled all DDlog profiling and update handler.
3. Changed the `apply_update` logic to send all updates to worker 0.

None of these changes made any difference: reducing the number of cores to 1 (when running with 4 workers) slows things down by a factor of 20.

Finally, I add one more change: close all input sessions on all workers, except worker 0, and now the benchmark runs 5x faster, which is 4x slower than on 4 cores.

Conclusion: There is no problem with this PR and I think it can be merged once we've addressed the TODOs. But I want to bring up the scheduling anomaly with @frankmcsherry.  I will try to create a minimal repro and submit an issue in timely.

**Comment by Kixiron** [comment]
That's really obvious now that I think about it. The `.step_while()` change came before I was more familiar with timely's internals, so I was thinking in a purely "what my program does" fashion and didn't know that timely did it's own unparking under the hood which lead to the thought of trying to avoid parking, since every park requires an unpark (which I assumed could only come from us). A really naive bug, that's my bad

**Comment by ryzhyk** [comment]
I wish I knew how to code without bugs. It's great that out tests caught it. Interestingly, the stream test was not even designed for this.

**Comment by ryzhyk** [comment]
I am going to close this PR and @Kixiron's original PR and create a fresh PR against the main DDlog repo, so I don't need to figure out merge conflicts.

**Pull request closed without merging**